Double click EmptyView to make a new document (and other minor quality of life changes)

AleksLitynski opened the pull request:
Feature changes:
- Can double click the empty view to open a new document
- Can double click a tab gutter (in the space where there isn't already a tab) to make a new document
- Can scroll the mouse wheel over a tab gutter to scroll tabs if more than 10 tabs are open

Technical changes:
- Moved EmptyView into a separate script. If a plugin ever wants to modify the EmptyView, this makes it possible.
- Light refactoring in rootview to break out some utility functions
- Added some helper functions to find out what part of the tab gutter a click was in

Risks:
- If any of my math is wrong, double clicks outside of the tab gutter could open new documents
- If the scrolling logic has a bug, it could eat scroll events that are meant to go deeper into the view hierarchy

Testing:
1. Build+Run lite xl
2. Close all tabs
3. Double click the background, confirm a tab is opened
4. Double click the tab gutter, confirm a tab is opened (you may have to edit the existing document first - if there's already a new document, lite xl doesn't want to make another new one)
5. Double click a tab and confirm a new tab isn't opened. Open 10+ tabs and double click the arrows left/right of the tabs, confirm a new tab isn't opened.
6. Create an h-split + v-split. Confirm the tab gutters in a vertical/horizontal split can be double clicked as expected.
7. Open 10+ tabs in all splits. Try scrolling over the tab bar. Confirm the tab bar scrolls.
8. Open a long document in each view. Confirm you can still scroll.
9. Try double clicking around other views (especially the tree view) and confirm you can't accidentally open more new docs.

Guldoman commented:
Oh this is nice. I'll check this out more in the weekend.

> Moved EmptyView into a separate script. If a plugin ever wants to modify the EmptyView, this makes it possible.

That's great and we should split even more.

I only gave this a quick look but it seems to be working fine. The only thing that bugs me is that the scroll direction is the opposite of what I expected.

Also I would set the mouse cursor to `arrow` while hovering the tab bar.

Wouldn't it be better to scroll through the tabs instead of just "panning" the tabs list?

adamharrison commented:
I like this too!

> Also I would set the mouse cursor to arrow while hovering the tab bar.

Definitely.

> Wouldn't it be better to scroll through the tabs instead of just "panning" the tabs list?

In terms of scrolling or panning, could we just turn this into a command for the keymap based on scroll? That way it's easy to rebind it to whichever you want, in whichever direction you wanted. (I also think it's backwards, though, from what I'd expect). 

That said, I think there's a bug with the panning/scrolling anyway. If I'm on tab 15, and I downscroll, it'll change me to the document for tab 14. However, if I upscroll, to go back to tab 15, it doesn't change me to tab 15. It seems like it only changes documents if the tab moves out of view.

IMO, it should be consistent, and pan, or scroll, but not both at the same time. I think a command for each would be appropriate that can then be bound to the keymap's wheelup or wheeldown with the appropriate predicate.

I'd vote for panning to be the default, though; but don't feel strongly about it.

Guldoman commented:
> In terms of scrolling or panning, could we just turn this into a command for the keymap based on scroll? That way it's easy to rebind it to whichever you want, in whichever direction you wanted. (I also think it's backwards, though, from what I'd expect).

This would be great.

> That said, I think there's a bug with the panning/scrolling anyway. If I'm on tab 15, and I downscroll, it'll change me to the document for tab 14. However, if I upscroll, to go back to tab 15, it doesn't change me to tab 15. It seems like it only changes documents if the tab moves out of view.

I think that's intended; the same behavior can be experienced by using the arrow buttons at the beginning and end of the tab bar.

> IMO, it should be consistent, and pan, or scroll, but not both at the same time. I think a command for each would be appropriate that can then be bound to the keymap's wheelup or wheeldown with the appropriate predicate.

Right now I don't think we can pan without having a visible selected tab.

adamharrison commented:
> I think that's intended; the same behavior can be experienced by using the arrow buttons at the beginning and end of the tab bar.

> Right now I don't think we can pan without having a visible selected tab.

Oh, interesting. To be honest, I have my tabs set to 1000, so I don't actually deal with scrolling (just personal preference). I don't really have a horse in this race; so if that's the way we want it to be, then that's fine.

Let's just be sure that we're ok with that behaviour. If we are, then sure; we're good. But IMHO if I'm panning, and not selecting a new tab, I'd assume that it wouldn't arbitrarily change scrolling through the list. But again, that's not my workflow, so I'll leave it in everyone else's capable hands, if they have other plans.

Guldoman commented:
> Let's just be sure that we're ok with that behaviour. If we are, then sure; we're good. But IMHO if I'm panning, and not selecting a new tab, I'd assume that it wouldn't arbitrarily change scrolling through the list.

Yeah, you're right, but it's a balance between that and not knowing where you left your selected tab. That's why I was suggesting just scrolling through the tabs.
I personally prefer having them show on multiple lines, but it's not in the scope of this PR (and might be a bit complicated, I'll have to check).

AleksLitynski commented:
I'm looking at a few other programs for reference:
- In firefox, if you open too many tabs, they overflow. If you scroll down, the tabs move towards the left (new tabs appearing on the right). The cursor never changes. The selected tab never changes.
- In chrome, the tabs never scroll. If you scroll the mouse, the selected tab changes. Down selects the next tab to the right, up selects the next tab to the left. If you open too many tabs, they literally overflow the end of the tab bar. You can scroll to them, but you can't see them (this seems like ugly behavior that chrome should handle better). The cursor never changes.
- In sublime text text, you can scroll if too many tabs are open. Scrolling down reveals more tabs on the right, like in firefox. Selection doesn't change, even if the selected tab scrolls off screen, like in firefox. The cursor never changes. Sublime does something cute where it piles up tab dividers on the left or right as tabs scroll off the edge. It gives you a sense that tabs are heaping up on the left/right.
- In vs code, scrolling down goes left, up goes right (just like firefox and sublime). Tabs can overflow. When you scroll, the selection is not changed. When the cursor is hovering over the tabs, it does change into a pointer finger (the cursor you get when you hover over a hyperlink). Uniquely, vs code shows a thin horizontal scroll bar below the tabs which only appears when you hover over the tabs. It can be click-and-dragged like a regular scroll bar.

Changes I'm planning to make:
1. The cursor will become the 'hyperlink' cursor when you hover over the tab bar
2. By default, scrolling down will reveal tabs on the right. Scrolling up will reveal tabs on the left (like in firefox). Scrolling will never change the active tab.
3. Users will be able to configure this behavior in two ways:
3a. They can override the scroll handling to flip the direction of scrolling
3b. They can override the handling so that the selected tab changes as they scroll

Guldoman commented:
> The cursor will become the 'hyperlink' cursor when you hover over the tab bar

I don't really like this. Would the cursor be 'hyperlink' while not hovering over a tab?

AleksLitynski commented:
I don't have any strong opinion about the cursor thing. Just let me know how you want it to behave and I'm happy to implement it.

Guldoman commented:
I think you could change how `RootView:on_mouse_moved` works, and request the arrow cursor when the mouse is over the tab bar. More or less like this:
```diff
@@ -1054,12 +1061,9 @@ function RootView:on_mouse_moved(x, y, dx, dy)
   self.overlapping_node = self.root_node:get_child_overlapping_point(x, y)
   
   local div = self.root_node:get_divider_overlapping_point(x, y)
-  local tab_index = self.overlapping_node and self.overlapping_node:get_tab_overlapping_point(x, y)
-  if self.overlapping_node and self.overlapping_node:get_scroll_button_index(x, y) then
-    core.request_cursor("arrow")
-  elseif div then
+  if div then
     core.request_cursor(div.type == "hsplit" and "sizeh" or "sizev")
-  elseif tab_index then
+  elseif self.overlapping_node and self.overlapping_node:in_tab_area(x, y) then
     core.request_cursor("arrow")
   elseif self.overlapping_node then
     core.request_cursor(self.overlapping_node.active_view.cursor)
```

Also I'd make it so that clicking on the tab area, the focus moves to that `Node`.
```diff
@@ -920,13 +925,15 @@ function RootView:on_mouse_pressed(button, x, y, clicks)
       return true
     end
   elseif not self.dragged_node then -- avoid sending on_mouse_pressed events when dragging tabs
-    local in_gutter = node:in_tab_area_gutter(x, y)
-    if in_gutter and button == "left" and clicks == 2 then
-      -- if the user double clicks inside the tab gutter, make a new document
-      command.perform "core:new-doc"
+    core.set_active_view(node.active_view)
+    local in_tab_area = node:in_tab_area(x, y)
+    if in_tab_area then
+      if button == "left" and clicks == 2 then
+        -- if the user double clicks inside the tab gutter, make a new document
+        command.perform "core:new-doc"
+      end
       return true
     end
-    core.set_active_view(node.active_view)
     if not self.on_view_mouse_pressed(button, x, y, clicks) then
       return node.active_view:on_mouse_pressed(button, x, y, clicks)
     end
```

This way the new doc command creates the new `DocView` in the right `Node`.

I noticed that `Node:get_scroll_button_index` returns `nil` then the number of `View`s is `1`, but I think it should be changes to return `nil` when `#self.views <= config.max_tabs`. I don't know if this could cause problems tho.

Can we also make it so that if the right scroll button isn't shown, we can double click there to open a new doc too?

Guldoman commented on data/core/commands/core.lua:
```diff
@@ -91,6 +91,8 @@ command.add(nil, {
   ["core:new-doc"] = function()
     core.root_view:open_doc(core.open_doc())
+    local node = core.root_view:get_active_node()
+    node:scroll_tabs_to_visible()
```
This should probably be done in `Node:set_active_view`.
If you do that, you'll also want to do something like:
```diff
@@ -19,15 +19,15 @@ function Node:new(type)
   self.size = { x = 0, y = 0 }
   self.views = {}
   self.divider = 0.5
-  if self.type == "leaf" then
-    self:add_view(EmptyView())
-  end
   self.hovered = {x = -1, y = -1 }
   self.hovered_close = 0
   self.tab_shift = 0
   self.tab_offset = 1
   self.tab_width = style.tab_width
   self.move_towards = View.move_towards
+  if self.type == "leaf" then
+    self:add_view(EmptyView())
+  end
 end
```
As calling `Node:get_visible_tabs_number` would result in trying to do math with `self.tab_offset` that is `nil`.

Guldoman commented on data/core/rootview.lua:
```diff
@@ -1057,12 +1057,11 @@ function RootView:on_mouse_moved(x, y, dx, dy)
   elseif div then
     core.request_cursor(div.type == "hsplit" and "sizeh" or "sizev")
-  elseif tab_index then
+  elseif self.overlapping_node and self.overlapping_node:in_tab_area(x, y) then
```
Would it be ok to merge `elseif self.overlapping_node and self.overlapping_node:in_tab_area(x, y) then` and `if self.overlapping_node and self.overlapping_node:get_scroll_button_index(x, y)`?

Guldoman reviewed (changes requested):
When we are panned to the last tabs and start closing them, the panning stays fixed while I would expect it to try and pan to show the remaining tabs.

As i realize i'm not making much sense, here's what i mean:
What I see after closing all the visible tabs but one:
![2021-12-02_22-26](https://user-images.githubusercontent.com/2798487/144505863-f239a4fe-5cff-4503-8231-c58c8013e7e8.png)
What I expect:
![2021-12-02_22-27](https://user-images.githubusercontent.com/2798487/144505885-0e31270f-65dd-49f3-a43f-dc72014a0a9f.png)

AleksLitynski commented:
I think I've addressed everything except `Also I'd make it so that clicking on the tab area, the focus moves to that Node.`, which I don't understand. It seems like when I click the tab area, the focus already moves, so maybe I fixed it accidentally, or I don't understand what you're suggesting.

Guldoman commented:
> I think I've addressed everything except `Also I'd make it so that clicking on the tab area, the focus moves to that Node.`, which I don't understand. It seems like when I click the tab area, the focus already moves, so maybe I fixed it accidentally, or I don't understand what you're suggesting.

Right now, if you click on an empty space of the tab bar, the click is sent to the `View` underneath, with the consequence that the caret moves. So yes, it receives focus, but for the wrong reason.

AleksLitynski commented:
I'm still a little shaky about the `Also I'd make it so that clicking on the tab area, the focus moves to that Node.` change, but I _think_ I did what you asked for. Let me know if there's anything else I can do here, otherwise, I'd love to get this pr merged.

Is there anywhere I should document how this works, in case people want to configure it in their init.lua file? Or is it on users to check the source code and figure it out themself for now?

Guldoman commented:
> I'm still a little shaky about the `Also I'd make it so that clicking on the tab area, the focus moves to that Node.` change, but I think I did what you asked for.

No, this is still broken. Here's a clip of what I mean:

https://user-images.githubusercontent.com/2798487/144733546-856c5554-36e7-4556-8321-3ba32a97e0df.mp4

Here I'm clicking on the tab bar, but the click gets sent to the `View` underneath. What I expect is that the `View` doesn't receive the click, but is set as the active view.

AleksLitynski commented:
I finally got back to this. December is a hectic month. Sorry it took so long.
- I merged the latest changes from master, which meant moving a lot of code over from RootView.lua to Node.lua.
- I fixed the 'clicking in the tab area moves the cursor' issue
- I fixed some other small issues I noticed. Namely
      - The scroll left/right buttons would only activate the nearest left/right tab IF the previous active tab had just scrolled off the edge. If you you the scroll wheel to move the previous active tab out of view, then clicked the left/right button, the active tab wouldn't change. Now, the left/right button always activate the left/right tab if the previously active tab isn't visible.
      - Scrolling tab bars wasn't 'sloppy' before. If you split your view into left and right areas, then focused the left area, then scrolled the right area's tab bar, the left tab bar would scroll. Now, the tab bar under the mouse will scroll, regardless of active view.

Anyways, let me know what else I can do on this!

Guldoman commented on data/core/emptyview.lua:
```diff
@@ -1,7 +1,11 @@
 
+-- This view is show in the background the first time lite xl is opened.
+-- Users also see it whenever they split the viewport and close all tabs
+-- in a single view.
```
The second part of this comment is not true anymore in 2.0.4 and in `master`, right?

Guldoman commented on data/core/rootview.lua:
```diff
@@ -131,7 +136,18 @@ function RootView:on_mouse_pressed(button, x, y, clicks)
-    return self.on_view_mouse_pressed(button, x, y, clicks) or node.active_view:on_mouse_pressed(button, x, y, clicks)
+    local in_gutter = node:in_tab_area_gutter(x, y)
+    if in_gutter and button == "left" then
+      if clicks == 2 then
```
What do you think about `clicks % 2 == 0`?
This way one could open multiple new docs without having to wait for the click timeout to expire.

In the end, it's also fine like this.

Guldoman reviewed (approved):
I really like this.
I'd just wait for someone else to review this too, as it's a change that touches some core components in important ways.

Guldoman commented:
> I finally got back to this. December is a hectic month. Sorry it took so long.

Don't worry about it, there's no hurry.

Thank you for your contribution!

adamharrison commented:
As a note; as per #pr-review-commitee resolution, this is closed as all the features present have been implemented over the past little while in other PRs. 

Sorry about that, @AleksLitynski ; we're finally get our act together with merging and following up on PRs, so really sorry about the yearish-long delay here.
